#### ACCQ203
# TP 10 - Invariants de similitude et LFSR

## Exo 145

In [31]:
MS = MatrixSpace(QQ,5,5)
K.<x> = PolynomialRing(QQ)

**1.**

In [3]:
M1 = MS([[-1841, -10363,   22304,  108021, -243809],
         [ 1366,   7695,  -16535,  -80130,  180869],
         [-1072,  -6088,   13069,   63408, -143144],
         [  506,   2951,   -6298,  -30700,   69343],
         [   82,    502,   -1061,   -5214,   11788]])

In [4]:
(x*matrix.identity(5)-M1).smith_form()[0]

[                      1                       0                       0                       0                       0]
[                      0                       1                       0                       0                       0]
[                      0                       0                       1                       0                       0]
[                      0                       0                       0          x^2 - 7*x + 10                       0]
[                      0                       0                       0                       0 x^3 - 4*x^2 - 11*x + 30]

Les invariants de similitude de $M_1$ sont donc $x^2 - 7x + 10$ et $x^3 - 4x^2 - 11x + 30$.

Son polynôme minimal est $x^3 - 4x^2 - 11x + 30$

On vérifie avec les fonctions de sage :

In [5]:
M1.minimal_polynomial()

x^3 - 4*x^2 - 11*x + 30

In [6]:
M1.characteristic_polynomial()

x^5 - 11*x^4 + 27*x^3 + 67*x^2 - 320*x + 300

In [7]:
M1.characteristic_polynomial() == (x^2 - 7*x + 10) * (x^3 - 4*x^2 - 11*x + 30)

True

**2.**

In [8]:
M2 = MS([[ 570,  1652,  -8251,  3807,   34007],
         [-178,  -522,   2666, -1196,  -10988],
         [ 540,  1573,  -7866,  3622,   32430],
         [ -42,  -118,    580,  -275,   -2387],
         [ 135,   393,  -1967,   905,    8109]])

In [9]:
(x*matrix.identity(5)-M2).smith_form()[0]

[                                           1                                            0                                            0                                            0                                            0]
[                                           0                                            1                                            0                                            0                                            0]
[                                           0                                            0                                            1                                            0                                            0]
[                                           0                                            0                                            0                                            1                                            0]
[                                           0                                            0  

Le seul invariant de similitude de $M_2$ est $x^5 - 16x^4 + 73x^3 - 18x^2 - 324x - 216$. Il s'agit aussi de son polynôme minimal et de son polynôme caractéristique.

In [10]:
M2.minimal_polynomial()

x^5 - 16*x^4 + 73*x^3 - 18*x^2 - 324*x - 216

In [11]:
M2.characteristic_polynomial()

x^5 - 16*x^4 + 73*x^3 - 18*x^2 - 324*x - 216

**3.**

In [12]:
M3 = MS([[ 64,  -300,    924,  -228,   3168],
         [-80,   404,  -1232,   304,  -4224],
         [ 35,  -175,    543,  -133,   1848],
         [-15,    75,   -231,    61,   -792],
         [-20,   100,   -308,    76,  -1052]])

In [13]:
(x*matrix.identity(5)-M3).smith_form()[0]

[             1              0              0              0              0]
[             0          x - 4              0              0              0]
[             0              0          x - 4              0              0]
[             0              0              0          x - 4              0]
[             0              0              0              0 x^2 - 8*x + 16]

Les invariants de similitude de $M_3$ sont $x-4$, $x-4$, $x-4$ et $x^2 - 8x + 16$.

Son polynome minimal est $x^2 - 8x + 16$.

In [14]:
M3.minimal_polynomial()

x^2 - 8*x + 16

In [15]:
M3.characteristic_polynomial()

x^5 - 20*x^4 + 160*x^3 - 640*x^2 + 1280*x - 1024

In [16]:
(x-4)^3 * (x^2 - 8*x + 16)

x^5 - 20*x^4 + 160*x^3 - 640*x^2 + 1280*x - 1024

## Exo 148

In [17]:
M1.rational_form()

[  0 -10|  0   0   0]
[  1   7|  0   0   0]
[-------+-----------]
[  0   0|  0   0 -30]
[  0   0|  1   0  11]
[  0   0|  0   1   4]

In [18]:
M2.rational_form()

[  0   0   0   0 216]
[  1   0   0   0 324]
[  0   1   0   0  18]
[  0   0   1   0 -73]
[  0   0   0   1  16]

In [19]:
M3.rational_form()

[  4|  0|  0|  0   0]
[---+---+---+-------]
[  0|  4|  0|  0   0]
[---+---+---+-------]
[  0|  0|  4|  0   0]
[---+---+---+-------]
[  0|  0|  0|  0 -16]
[  0|  0|  0|  1   8]

On retrouve bien les matrices compagnons des invariants de similitudes des différents opérateurs.

## Exo 153

**1.** Pour rappel, les invariants de similitude de $M_1$ sont donc $x^2 - 7x + 10$ et $x^3 - 4x^2 - 11x + 30$. On détermine leur factorisation :

In [23]:
(x^2 - 7*x + 10).factor()

(x - 5) * (x - 2)

In [24]:
(x^3 - 4*x^2 - 11*x + 30)

x^3 - 4*x^2 - 11*x + 30

D'où la forme de Jordan de $M_1$ :

In [20]:
M1.jordan_form()

[-3| 0| 0| 0| 0]
[--+--+--+--+--]
[ 0| 5| 0| 0| 0]
[--+--+--+--+--]
[ 0| 0| 5| 0| 0]
[--+--+--+--+--]
[ 0| 0| 0| 2| 0]
[--+--+--+--+--]
[ 0| 0| 0| 0| 2]

**2.** Le seul invariant de similitude de $M_2$ est $x^5 - 16x^4 + 73x^3 - 18x^2 - 324x - 216$

In [26]:
(x^5 - 16*x^4 + 73*x^3 - 18*x^2 - 324*x - 216).factor()

(x + 1)^2 * (x - 6)^3

In [28]:
M2.jordan_form()

[-1  1| 0  0  0]
[ 0 -1| 0  0  0]
[-----+--------]
[ 0  0| 6  1  0]
[ 0  0| 0  6  1]
[ 0  0| 0  0  6]

**3.** Les invariants de similitude de $M_3$ sont $x-4$, $x-4$, $x-4$ et $x^2 - 8x + 16$.

In [32]:
(x^2-8*x+16).factor()

(x - 4)^2

In [30]:
M3.jordan_form()

[4 1|0|0|0]
[0 4|0|0|0]
[---+-+-+-]
[0 0|4|0|0]
[---+-+-+-]
[0 0|0|4|0]
[---+-+-+-]
[0 0|0|0|4]

## Exo 160

**1.** Comme borne pour l'exposant du groupe multiplicatif $ {\mathbb F}_q[x]/ \langle \chi(x) \rangle $, on peut choisir $q^n - 1$.

**2.**

In [33]:
def periodePoly(p):

    P = p.parent()
    q = P.characteristic()
    n = p.degree()

    Pmodp.<x> = P.quotient(p)

    m = int(sqrt(q^n-1))+1
    T = []
    e = 1

    for j in range(m):
        T += [e]
        e = e*x
        if (j>0 and e==1):
            return j+1

    h = x^(-m)
    gamma = h

    for i in range(m):
        for j in range(m):
            if gamma == T[j]:
                return (i+1)*m+j
        gamma = gamma*h

**3.**

In [46]:
F2 = FiniteField(2)
P.<x> = PolynomialRing(F2)
chi1 = 1 + x + x^2 + x^3 + x^4
chi2 = 1 + x + x^2 + x^4
chi3 = 1 + x^3 + x^6

In [47]:
periodePoly(chi1)

5

In [48]:
periodePoly(chi2)

7

In [49]:
periodePoly(chi3)

9

In [50]:
F5 = FiniteField(5)
P.<x> = PolynomialRing(F5)
chi4 = 3 + x^2
chi5 = 3 + 3*x + x^2

In [51]:
periodePoly(chi4)

8

In [52]:
periodePoly(chi5)

24

**4.**

In [34]:
def periodePoly_naif(p):
    t = 1
    while not(p.divides(x^t-1)):
        t += 1
    return t

In [57]:
timeit("periodePoly(chi5)")

125 loops, best of 3: 2.26 ms per loop


In [58]:
timeit("periodePoly_naif(chi5)")

625 loops, best of 3: 142 µs per loop


L'algorithme naïf est plus rapide que l'algorithme Baby Step, Giant Step sur cet exemple. Pour cause, l'exemple est de petit degré et l'algo utilise une structure de donnée plus compliquée.

## Exo 161

In [61]:
def periode(C,v0):
    n = len(v0)
    v = v0
    for r in range(1,100):
        v = C*v
        if v == v0:
            break
    return r

In [62]:
C=matrix(F2,[[0,1,0,0],[0,0,1,0],[0,0,0,1],[1,1,1,1]])
V = list(cartesian_product_iterator([F2]*4))
t = []
for v in V:
    t += [periode(C,vector(v))]
t

[1, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5]

In [63]:
C=matrix(F2,[[0,1,0,0],[0,0,1,0],[0,0,0,1],[1,1,1,0]])
V = list(cartesian_product_iterator([F2]*4))
t = []
for v in V:
    t += [periode(C,vector(v))]
t

[1, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 1]

In [64]:
C=matrix(F2,[[0,1,0,0,0,0],[0,0,1,0,0,0],[0,0,0,1,0,0],[0,0,0,0,1,0],[0,0,0,0,0,1],[1,0,0,1,0,0]])
V = list(cartesian_product_iterator([F2]*6))
t = []
for v in V:
    t += [periode(C,vector(v))]
t

[1,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9]

In [65]:
C=matrix(F5,[[0,1],[2,0]])
V = list(cartesian_product_iterator([F5]*2))
t = []
for v in V:
    t += [periode(C,vector(v))]
t

[1, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8]

In [66]:
C=matrix(F5,[[0,1],[2,2]])
V = list(cartesian_product_iterator([F5]*2))
t = []
for v in V:
    t += [periode(C,vector(v))]
t

[1,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24,
 24]

## Exo 166

In [67]:
def myBerlekampMassey(r,q):
    n = len(r)
    Fq = GF(q)
    P.<x> = PolynomialRing(Fq)
    c = P(1)
    l = 0
    cStar = P(1)
    dStar = Fq(1)
    m = -1
    for k in range(n):
        d = Fq( r[k] + sum([c[i]*r[k-i] for i in range(1,l+1)]) )
        if d!=0 :
            t = c
            c = c - d * (dStar)^(-1) * cStar * x^(k-m)
            if (l<=k/2):
                l = k+1-l
                cStar = t
                dStar = d
                m = k
    return c,l

In [68]:
myBerlekampMassey([1,0,1,1,0,0,0,0,1,1,0,1,1,0,0,0,0,1,1,0,1,1,0,0,0,0,1,1,0,1,1,0,0,0,0,1],2)

(x^6 + x^3 + 1, 6)